ahnlich/ai/src/engine/ai/providers/ort.rs

```diff
@@ -0,0 +1,266 @@
+
+use std::convert::TryFrom;
+use std::default::Default;
+use std::fmt;
+use std::path::{Path, PathBuf};
+use std::thread::available_parallelism;
+
+#[derive(Default)]
+pub struct ORTProvider {
+    cache_location: Option<PathBuf>,
+    cache_location_extension: PathBuf,
+    supported_models: Option<SupportedModels>,
+    model: Option<ORTModel>,
+}
+
+impl fmt::Debug for ORTProvider {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ORTProvider")
+            .field("cache_location", &self.cache_location)
+            .field("cache_location_extension", &self.cache_location_extension)
+            .field("supported_models", &self.supported_models)
+            .finish()
+    }
+}
+
+#[derive(Default)]
+pub struct ORTImageModel {
+    repo_name: String,
+    weights_file: String,
+    session: Option<Session>,
+    input_param: String,
+    output_param: String,
+}
+
+pub enum ORTModel {
+    Image(ORTImageModel),
+}
+
+impl TryFrom<&SupportedModels> for ORTModel {
+    type Error = AIProxyError;
+
+    fn try_from(model: &SupportedModels) -> Result<Self, Self::Error> {
+        let model_type = match model {
+            SupportedModels::Resnet50 => Ok(ORTImageModel {
+                repo_name: "Qdrant/resnet50-onnx".to_string(),
+                weights_file: "model.onnx".to_string(),
+                input_param: "input".to_string(),
+                output_param: "image_embeds".to_string(),
+                ..Default::default()
+            }),
+            SupportedModels::ClipVitB32 => Ok(ORTImageModel {
+                repo_name: "Qdrant/clip-ViT-B-32-vision".to_string(),
+                weights_file: "model.onnx".to_string(),
+                input_param: "pixel_values".to_string(),
+                output_param: "image_embeds".to_string(),
+                ..Default::default()
+            }),
+            _ => Err(AIProxyError::AIModelNotSupported),
+        };
+        Ok(ORTModel::Image(model_type?))
+    }
+}
+
+impl ORTProvider {
+    pub(crate) fn new() -> Self {
+        Self {
+            cache_location: None,
+            cache_location_extension: PathBuf::from("huggingface"),
+            supported_models: None,
+            model: None,
+        }
+    }
+
+    pub fn normalize(v: &[f32]) -> Vec<f32> {
+        let norm = (v.iter().map(|val| val * val).sum::<f32>()).sqrt();
+        let epsilon = 1e-12;
+
+        // We add the super-small epsilon to avoid dividing by zero
+        v.iter().map(|&val| val / (norm + epsilon)).collect()
+    }
+}
+
+impl ProviderTrait for ORTProvider {
+    fn set_cache_location(&mut self, location: &Path) -> &mut Self {
+        self.cache_location = Some(location.join(self.cache_location_extension.clone()));
+        self
+    }
+
+    fn set_model(&mut self, model: &SupportedModels) -> &mut Self {
+        self.supported_models = Some(*model);
+        self
+    }
+
+    fn load_model(&mut self) -> &mut Self {
```

Comment:
`load_model` looks very fragile as there's a lot of unwraps.... can we instead pipe these upwards?

Reply:
Valid point. I agree.